sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/grouping.scala

```diff
@@ -231,3 +235,32 @@ object GroupingID {
     if (SQLConf.get.integerGroupingIdEnabled) IntegerType else LongType
   }
 }
+
+object GroupingAnalytics {
+  def unapply(exprs: Seq[Expression])
+  : Option[(Seq[Seq[Expression]], Seq[Expression])] = {
+    val resolved = exprs.map {
+      case gs: BaseGroupingSets => gs.childrenResolved
+      case other => other.resolved
+    }.forall(_ == true)
+    if (!resolved) {
+      None
+    } else if (!exprs.exists(e => e.find(_.isInstanceOf[BaseGroupingSets]).isDefined)) {
```

Comment:
do we need to call `find`? I think `BaseGroupingSets` can only appear in the top level.

Reply:
BTW this check can go first, as `isInstanceOf[BaseGroupingSets]` is cheaper to run

Reply:
DOne